### PR TITLE
nixos/boot: add /var/lib/nixos to pathsNeededForBoot

### PR DESCRIPTION
**Pull request by deviant**
/var/lib/nixos is used by update-users-groups.pl in the activation
script for storing uid/gid mappings. If this has its own mountpoint
(as is the case in some setups with fine-grained bind mounts pointing
into persistent storage), the mappings are written to /var/lib, /var,
or /. These may be backed by a tmpfs or (otherwise ephemeral storage),
resulting in the mappings not persisting between reboots.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
I just built a system and `/var/lib/nixos` is on its own `zfs` partition. I don't think I have rebooted this system yet. Should I expect any problems when I do? 🤔

**Comment by deviant** [comment]
@aanderse If `/var/lib` isn't persistent through boots, you should expect the UID mapping to be reset on reboot. Unless you have manually marked it as `neededForBoot`, it should be mounted after the activation script is run. Testing this shouldn't be difficult, just change some of the values in `/var/lib/nixos/uid-map` (and ensure `/etc/passwd` is not persistent), or add a new user which gets sorted (by name, alphabetically) before an existing one.

**Comment by aanderse** [comment]
@deviant thanks for that. I guess this should have been obvious to me, but honestly I'm sure I would have been confused the first time I rebooted that machine and had some permissions issues :laughing: Fortunately many services use `StateDirectory` which would save the day for a number of systems.

A big :+1: on this PR from me.

**Comment by deviant** [comment]
@aanderse I ran into it when migrating an existing system to have an ephemeral root filesystem. Wondering why I was logged into a different user's home directory yet still had the ability to use sudo was a little alarming, before I realised what had happened. If you're setting up a fresh system, you would never run into this problem, so I totally understand. (:

**Comment by aanderse** [comment]
Informational ping to @grahamc because of [erase your darlings](https://grahamc.com/blog/erase-your-darlings).

**Comment by deviant** [comment]
@aanderse ping?

**Comment by aanderse** [comment]
Right, sorry. I was hoping @Mic92 would double check this before review. @andir, @jonringer, @flokli, etc... would all be great to double check this as well.

Would be very nice to have before we freeze for `21.05`...

If no one replies by the weekend ping me again and we merge. This looks pretty straight forward and I've been running it on my system for a bit.

**Comment by andir** [comment]
This looks fine but if it is a feature we support there should be a test for it.

**Comment by deviant** [comment]
Tagging @talyz because of https://github.com/nix-community/impermanence — any other directories you're aware of? I'm definitely not certain this is the last of them, I just haven't run into any others yet.

@andir by "feature" do you mean booting from a fresh root partition? This seems to be growing in popularity & has a number of blog posts written about at this point, so I assumed it was, implicitly

**Comment by talyz** [comment]
This looks good to me!

I'm not aware of any other paths that would need to be added to the list, and since I always use `users.mutableUsers = false`, I wasn't aware of this one either. :)

**Comment by deviant** [comment]
I did have `mutableUsers = false`, was only ticked off to this because I was also migrating a disk from one machine to another (and thus users were created in a different order than the UIDs were allocated in its filesystem). I suppose I could have hardcoded UIDs to avoid this, but that shouldn't be necessary if there's state.

**Comment by deviant** [comment]
> If no one replies by the weekend ping me again and we merge. This looks pretty straight forward and I've been running it on my system for a bit.

@aanderse AFAICT nobody has raised concerns with this, want to merge now?

**Comment by andir** [comment]
> @andir by "feature" do you mean booting from a fresh root partition? This seems to be growing in popularity & has a number of blog posts written about at this point, so I assumed it was, implicitly

Yes. Also booting NixOS on a fresh partition works. It just doesn't behave as expected in the case that `/var/lib/nixos` (an implementation detail of NixOS) is on a different partition. In this case something that is an implementation detail has leaked into the users configuration.
From my current perspective it is a bug in `nix-community/impermanence` that should be fixed there. If you mutate the systems mount table in a way that isn't intended you are basically on your own.

*But* since it is very likely that most/all of NixOS users (and certainly me) can relate to the desire of having an as pure as possible system we can and **should also add a fix to NixOS**.

I am not yet decided if, from NixOS perspective, it is fair to assume that `/var/lib` is sufficient. it is kind of a philosophical conversation to have.

My initial hesitation comes due to the chance of "feature creep" in NixOS. We should only carry "features" that we actually know are there and that are reassured to be working by having them tested in some way or another.

The proper fix for this is probably checking if `/var/lib/nixos` has a mountpoint and only then requiring it for system boot. This is essential the same fix as the one you proposed but makes that very case specific. It opens the opportunity to properly document that case and also ensures that systems (without that change) don't carry those modifications. In this case it is basically bike shedding about which fix to use but if this were to be a larger fix I'd bring up the same argument and approach.


tl;dr: Your fix is fine and I am not opposed to it. It would just be better if we wouldn't have to care about implementation details leaking and/or if we could programmatically enforce that this doesn't happen again (with another path).

**Comment by deviant** [comment]
> Yes. Also booting NixOS on a fresh partition works. It just doesn't behave as expected in the case that `/var/lib/nixos` (an implementation detail of NixOS) is on a different partition. In this case something that is an implementation detail has leaked into the users configuration.
> From my current perspective it is a bug in `nix-community/impermanence` that should be fixed there. If you mutate the systems mount table in a way that isn't intended you are basically on your own.

Ah, I feel I misexplained: I put `/var/lib/nixos` on its own ZFS filesystem due to splitting up the various directories in `/var/lib` between ephemeral and persistent filesystems, as suggested in https://grahamc.com/blog/nixos-on-zfs. This isn't related to impermanence and would have occurred had the root filesystem been a tmpfs, etc as well. Most other people who have done this have probably only put `/var/lib` or just `/var` on its own filesystem (or as a subdirectory of a state filesystem, etc), and thus would not have run into this issue.

> I am not yet decided if, from NixOS perspective, it is fair to assume that `/var/lib` is sufficient. it is kind of a philosophical conversation to have.

Given that `/var/lib/nixos` is used at boot, and NixOS shouldn't need to know whether it has a mountpoint of its own or not, I'm inclined to say yes. If we put systemd in the initrd, we'll probably have a unit for this regardless, IIUC.

> My initial hesitation comes due to the chance of "feature creep" in NixOS. We should only carry "features" that we actually know are there and that are reassured to be working by having them tested in some way or another.

I feel that them being advocated and written about by several of the core contributors counts as a feature?

> The proper fix for this is probably checking if `/var/lib/nixos` has a mountpoint and only then requiring it for system boot. This is essential the same fix as the one you proposed but makes that very case specific. It opens the opportunity to properly document that case and also ensures that systems (without that change) don't carry those modifications. In this case it is basically bike shedding about which fix to use but if this were to be a larger fix I'd bring up the same argument and approach.

Well, it's a bit of a hardcoded approach to even put this all in one place. Having a more structured approach would be nice (and I think this would come with systemd, as I mentioned above?). I think I'd disagree about predicating this; it's already a no-op if there's no filesystem declared for it (if you have a root partition only, for instance, all of the other directories are implicitly present, due to their hierarchical nature), which makes this really into a list of directories used at boot, not filesystems. We could also split this list into one line each, and add a comment for every directory which is specifically depended on, but the blame should make it evident where this came from at the very least.

> tl;dr: Your fix is fine and I am not opposed to it. It would just be better if we wouldn't have to care about implementation details leaking and/or if we could programmatically enforce that this doesn't happen again (with another path).

I agree. I'd definitely prefer if there were a structured option which was set at the location of use, rather than what we currently have.

**Comment by aanderse** [comment]
Any update on this one?

**Comment by deviant** [comment]
This seems like a trivial change that would avoid subtle breakage for users with heavily partitioned systems, will change nothing for users who don't, and is unlikely to introduce any regressions. It will additionally make the list of mountpoints we depend on during boot more explicit. Is there a good reason to not just accept it as-is, given the lack of foreseeable downsides?

**Review by talyz — APPROVED** [review]
I don't see any reason not to merge this either: it's a standard path which is required for boot, so it makes sense to add it to `pathsNeededForBoot`. It also shouldn't impact people who don't keep `/var/lib/nixos` on its own partition negatively in any way.
